### PR TITLE
fix: Skip building functions/layers if codeuri is a zip file

### PR DESCRIPTION
**Pull request by aahung**
#### Which issue(s) does this change fix?

#2703

#### Why is this change necessary?

currently (tested with 1.10.0, 1.19.0 and 1.21.1), when codeuri is a zip file, `sam build` will try to build using the zip file's path as source directory and `LambdaBuilder` will fail with error like this:
```
uilding codeuri: Archive.zip runtime: python3.7 metadata: {} functions: ['Function2']
Running PythonPipBuilder:ResolveDependencies

Build Failed
Error: PythonPipBuilder:ResolveDependencies - Requirements file not found: /Users/xinhol/Downloads/test-20210317/Archive.zip/requirements.txt
``` 

Some debugging context:

image

This issue becomes more impacting because before 1.20.0 was released, customers need to build functions in nested stacks themselves. Setting `codeuri` to a artifact zip was one of the practices to package/deploy nested stacks. With 1.20.0, `sam build` will try to build those zip files, which results in build error. With this patch, zip files will not be built again.

#### How does it address the issue?

Skip building functions or layers when their codeuri is a zip file. 
**It is excluded in BuildContext instead of providers, because we still want these functions to be local invoke-able.**

#### What side effects does this change have?

no side effect.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/build/build_context.py [hunk 1]
```diff
@@ -265,3 +265,27 @@ def _collect_single_buildable_layer(
...
         resource_collector.add_layer(layer)
+
+    @staticmethod
+    def _is_function_buildable(function: Function):
+        # no need to build inline functions
+        if function.inlinecode:
+            LOG.debug("Skip building inline function: %s", function.full_path)
+            return False
+        # no need to build functions that are already packaged as a zip file
+        if isinstance(function.codeuri, str) and function.codeuri.endswith(".zip"):
```
Comment (mndeveci) [inline samcli/commands/build/build_context.py:276]: Are we going to consider an S3 location as well?

Comment (aahung) [reply samcli/commands/build/build_context.py:276]: I am still checking details why we handled the S3 this way (this code exists before our git history) with @jfuss. I am gonna handle s3 in a separate PR

```py
        codeuri = resource_properties.get(code_property_key, SamBaseProvider.DEFAULT_CODEURI)
        # CodeUri can be a dictionary of S3 Bucket/Key or a S3 URI, neither of which are supported
        if isinstance(codeuri, dict) or (isinstance(codeuri, str) and codeuri.startswith("s3://")):
            if not ignore_code_extraction_warnings:
                LOG.warning(
                    "Lambda function '%s' has specified S3 location for CodeUri which is unsupported. "
                    "Using default value of '%s' instead",
                    name,
                    SamBaseProvider.DEFAULT_CODEURI,
                )
            return SamBaseProvider.DEFAULT_CODEURI
        return cast(str, codeuri)
```

**Review by sriram-mv — COMMENTED** [review]
Overall, this looks good, should we add an integ test as well?

**Review by sriram-mv — APPROVED** [review]

File: samcli/commands/build/build_context.py (hunk 1 above)
Comment (qingchm) [inline samcli/commands/build/build_context.py:276]: should we add a strip() before checking the endswith?

File: samcli/commands/build/build_context.py [hunk 2]
```diff
@@ -265,3 +265,27 @@ def _collect_single_buildable_layer(
...
+        return True
+
+    @staticmethod
+    def _is_layer_buildable(layer: LayerVersion):
+        # if build method is not specified, it is not buildable
+        if not layer.build_method:
+            LOG.debug("Skip building layer without a build method: %s", layer.full_path)
+            return False
+        # no need to build layers that are already packaged as a zip file
+        if isinstance(layer.codeuri, str) and layer.codeuri.endswith(".zip"):
```
Comment (qingchm) [inline samcli/commands/build/build_context.py:288]: Same question here

File: samcli/commands/build/build_context.py (hunk 1 above)
Comment (aahung) [reply samcli/commands/build/build_context.py:276]: the `codeuri` is from yaml, so we don't need to.

**Review by qingchm — APPROVED** [review]

**Pull request merged**